pkg/network/mycelium/mycelium.go

```diff
@@ -0,0 +1,235 @@
+	if s.cfg.PublicKey == "" {
+		panic("EncryptionPublicKey empty")
+	}
+
+	return hex.DecodeString(s.cfg.PublicKey)
+}
+
+func (s *MyServer) InspectMycelium() (inspection MyceliumInspection, err error) {
+	// we check if the file exists before we do inspect because mycelium will create a random seed
+	// file if file does not exist
+	_, err = os.Stat(confPath)
+	if err != nil {
+		return inspection, err
+	}
+
+	bin, err := exec.LookPath(myBin)
+	if err != nil {
+		return inspection, err
+	}
+
+	cmd := fmt.Sprintf(`exec ip netns exec %s %s inspect --json --key-file %s `, s.ns, bin, confPath)
+	output, err := exec.Command("sh", "-c", cmd).Output()
+	if err != nil {
+		return inspection, errors.Wrap(err, "failed to inspect mycelium ip")
+	}
+
+	if err := json.Unmarshal(output, &inspection); err != nil {
+		return inspection, errors.Wrap(err, "failed to load mycelium information from key")
+	}
+
+	return
+}
+
+// IP return the address in the 400::/7 subnet allocated by mycelium
+func (m *MyceliumInspection) IP() net.IP {
+	return net.IP(m.Address)
+}
+
+// Subnet return the 400::/64 subnet allocated by mycelium
+func (m *MyceliumInspection) Subnet() (subnet net.IPNet, err error) {
+	ipv6 := m.Address.To16()
+	if ipv6 == nil {
+		return subnet, errors.Errorf("invalid mycelium ip")
+	}
+
+	ip := make(net.IP, net.IPv6len)
+	copy(ip[0:8], ipv6[0:8])
+
+	subnet = net.IPNet{
+		IP:   ip,
+		Mask: net.CIDRMask(64, 128),
+	}
+
+	return
+}
+
+// Gateway derive the gateway IP from the mycelium IP in the /64 range.
+func (m *MyceliumInspection) Gateway() (gw net.IPNet, err error) {
+	subnet, err := m.Subnet()
+	if err != nil {
+		return gw, err
+	}
+
+	ip := subnet.IP
+	ip[net.IPv6len-1] = 1
+
+	gw = net.IPNet{
+		IP:   ip,
+		Mask: net.CIDRMask(64, 128),
+	}
+
+	return
+}
+
+// Tun return the name of the TUN interface created by mycelium
+func (s *MyServer) Tun() string {
+	return s.cfg.TunName
+}
+
+// SubnetFor return an IP address out of the node allocated subnet by hasing b and using it
+// to generate the last 64 bits of the IPV6 address
+func (m *MyceliumInspection) SubnetFor(b []byte) (net.IPNet, error) {
+	subnet, err := m.Subnet()
+	if err != nil {
+		return net.IPNet{}, err
+	}
+
+	ip := make([]byte, net.IPv6len)
+	copy(ip, subnet.IP)
+
+	subIP, err := subnetFor(ip, b)
+	if err != nil {
+		return net.IPNet{}, err
+	}
+
+	return net.IPNet{
+		IP:   subIP,
+		Mask: net.CIDRMask(64, 128),
+	}, nil
+}
+
+func subnetFor(prefix net.IP, b []byte) (net.IP, error) {
+	h := sha512.New()
+	if _, err := h.Write(b); err != nil {
+		return nil, err
+	}
+	digest := h.Sum(nil)
+	copy(prefix[8:], digest[:8])
+	return prefix, nil
+}
+
+func writeKey(path string, privateKey ed25519.PrivateKey) error {
+	if err := os.MkdirAll(filepath.Dir(path), 0770); err != nil {
+		return err
+	}
+
+	f, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	key := x25519.EdPrivateKeyToX25519([]byte(privateKey))
+
+	return json.NewEncoder(f).Encode(key)
```

Comment:
Why are u using `json` here. the priv key of mycelium is not json encoded. It's a binary file. i don't believe this file is loaded by mycelium and is either ignored is overridden.

If this ever worked, then i assume it's because mycelium just read key length bytes from the file and used that as seed for the key!

This is definitely an issue, because it means the key u generated here is not secure. because it's not the full "Random" key but only part of the key json encoded bytes.

Do you get what I mean ?